Add gpg self-signature verification support

lukpueh opened the pull request:
*I apologize for the size of this PR, but separating the commits into more (inter-dependent) PRs would only increase the PR handling effort and still leave at least one PR with a large diff. Let me know if there is anything I can do to facilitate code review. A guide to this PR is provided below.*

**Fixes issue #**:
- Depends on PR #255, please review and merge there first (ed1ccd5 - 050b7cb)
- Provides infrastructure required by #245

**Description of the changes being introduced by the pull request**:
This PR adds signature verification on gpg public key export for two types of self-signatures:
- Self-certificates related to the primary key, which can be used to acquire additional certified information about the the primary key, such as the key expiration date.
- Subkey binding signatures, that bind each subkey to the primary key. So far we just assumed that a subkey attached to a primary key belongs to the primary key without verifying the binding signature. This PR discards all subkey that aren't verifiably bound to the primary key.

The lion share of this PR is provided by d3b6c44, which updates gpg public key parsing  to verify the corresponding self-signatures. 
Suggestions to re-organize the relevant functions, i.e. `get_pubkey_bundle `, `parse_pubkey_bundle`, `_assign_certified_key_info` and `_get_verified_subkeys`, are appreciated.

So far we only supported gpg binary type signatures (i.e. signatures over in-toto metadata) and restricted signature parsing accordingly. b7ce963 updates signature parsing to support the self-signature types described above. 

For binary type signature creation and verification we only support SHA256. 713929b, 9388560 add changes to support SHA1, and SHA512 too (for self-signatures only!), which, based on real-word keyring inspections, seem to be widely-used.  

4181469 updates signature parsing to return additional information, which is needed for further signature handling (as part of this PR), and also to get additional information about the primary key, such as the expiration date (required for #245). Note that this PR only provides certificate verification support, while the information extraction should be handled in #245 (see [inline `# TODO:` comment](https://github.com/lukpueh/in-toto/blob/d3b6c44fe76013568f3459f67ba2ed35db3935f0/in_toto/gpg/common.py#L342-L343)).

Further tangential clean-up/refactor commits include:
-  26c1fbf to remove an obsolete `else` clause
- ae4f5d0 to enhance some related error messages and fix comments and docstrings 
-  4d825be to slightly modify signature keyid handling

**Caveats/TODOs**
- Lacks proper testing (I'm inclined to do this in a separate PR out of sympathies for the reviewers)
- Does not handle contradictory self-certificates (see [inline `# TODO:` comment](https://github.com/lukpueh/in-toto/blob/d3b6c44fe76013568f3459f67ba2ed35db3935f0/in_toto/gpg/common.py#L343-L346), should be addressed with #245)
- Does not handle key revocation certificates (see https://github.com/in-toto/in-toto/issues/263)


**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Add partial SHA1 and SHA512 gpg signature support

Support verification of gpg signatures that use SHA1 and SHA512.
Together with SHA256, these seem to be the most widely used
hashing algorithms for signatures over keys, such as
self-signatures.
Use the following command to see the distribution of hash
algos on your local gpg keyring (2, 8 and 10 should stick out).

gpg --export | gpg --list-packets | grep ":signature packet:" -a2 |\
grep -o "digest algo [0-9]*" | sort | uniq -c

See RFC4880 9.4 for the list of all hash algorithm IDs.

We continue to use SHA256 exclusively for metadata signing
and verification. This addition is only needed to support
self-signature verification.

Commit:
Update supported hash algos in signature parsing

Add optional supported_hash_algorithms argument to
parse_signature_packet to support parsing signature packets
with other hash algos. So far only SHA256 was supported, which
remains default.

Commit:
Update supported types in signature parsing

Add optional supported_signature_types argument to
parse_signature_packet to support parsing signature packets
of other types, e.g. different types of self-signatures.
So far only SIGNATURE_TYPE_BINARY packets were supported, which
remains the default.

Commit:
Enhance gpg error messages and fix comments/docs

Commit:
Add optional info to signature

Update parse_signature_packet to add an optional opaque info
dictionary to the returned signature. Whether it is added or
not may be configured via boolean "include_info" parameter. Per
default it is not added.

Currently the info dictionary contains the signature_type and
hash_algorithm and an empty list that may be used to add
subpackets such as a key expiration date.
The corresponding subpacket parsing code is already there
(commented out).

Commit:
Update keyid handling in parse_signature_packet

Update parse_signature_packet to only return the optional
short_keyid if it was parsed from the signature packet to (better)
match in_toto.gpg.formats.SIGNATURE_SCHEMA.

Note that the returned signature still might fail the schema check,
due to an empty keyid, but it is the responsibility of the caller
to adjust accordingly.

This commit also updates comments and calling functions.

Commit:
Remove obsolete else clause in parse_pubkey_payload

Commit:
Add gpg self-certificate verification support

This commit modifies the gpg public key parsing function to parse
additional packets, including User ID and User Attribute packets,
as well as signature packets corresponding to any of the other
parsed packets.

Additionally, the commit adds a helper function to verify
self-certificates that may carry information about the primary key,
such as key expiration date.

And it adds a helper function to verify sub key binding signatures
and updates the public key export function to only return subkeys
that are verifiably bound to the primary key.

aaaaalbert commented on in_toto/gpg/util.py:
```diff
@@ -247,3 +310,37 @@ def is_version_fully_supported():
+    raise ValueError("Hash algorithm '{}' not supported, must be one of '{}' "
+        "(see RFC4880 9.4. Hash Algorithms).".format(hash_algorithm_id,
+        {in_toto.gpg.constants.SHA1, in_toto.gpg.constants.SHA256,
+         in_toto.gpg.constants.SHA512}))
```
If this is never extended to support other hash types, my comment is moot, but there's a perhaps more concise way that avoids the `if` cascade:
```python
  supported_hashing_algorithms = [in_toto.gpg.constants.SHA1,
      in_toto.gpg.constants.SHA256, in_toto.gpg.constants.SHA512]

  corresponding_hashing_classes = [hashing.SHA1, hashing.SHA256, hashing.SHA512]

  hashing_class = dict(zip(supported_hashing_algorithms, corresponding_hashing_classes))

  try:
    return hashing_class[hash_algorihm_id]
  except KeyError:
    raise ValueError("Hash algorithm '{}' not supported, must be one of '{}' "
        "(see RFC4880 9.4. Hash Algorithms).".format(hash_algorithm_id,
        supported_hashing_algorithms))
```

(And had 938856026d091878883db9f837fe762dceaa4c03 not dropped `SUPPORTED_HASH_ALGORITHMS` from `gpg/constants.py`, the proposed code could be even DRYer.)

lukpueh replied:
I opted for the if/else variant because it seemed a tad easier to read, and I am not sure if we need to support other algos in the near future. But your approach is definitely more elegant.

SantiagoTorres replied:
I was going to suggest something similar. I'm fine either way though.

aaaaalbert commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+      elif packet_type == PACKET_TYPE_PRIMARY_KEY and \
+          key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("Unexpected primary key.")
+
```
I always find `if` cascades with complex causes difficult to read, as I'm never sure if I'm not missing a combination of conditions that's not handled. (I do think however that L193-L251 make sense.)

The `elif` in L198-199 could be rolled into L205+ as a sub-`if` clause.

lukpueh replied:
Agreed, that's why I added comprehensive comments.

> The `elif` in L198-199 could be rolled into L205+ as a sub-`if` clause.

That's how I wrote it initially, but then I thought it would be more readable the way it is now, as it is less arrow-head'ish.

aaaaalbert replied:
> arrow-head'ish

Hmm, like, *two* levels in doesn't feel *that* spiky to me :-)

lukpueh replied:
It would be 4 levels already: `while`, `try`, `elif`, `if` ... 5 if you count the `def`... :P

aaaaalbert replied:
...which is *technically* correct, but I think the `def` and `while` and `try` only count as overarching brackets. So: two ;-)

lukpueh replied:
Agreed, it's only two (...technically). I'll still keep this one.

aaaaalbert commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+      # - disallow duplicate packets
+      if packet_type != PACKET_TYPE_PRIMARY_KEY and \
+          not key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("First packet must be a primary key ('{}'), got '{}'."
```
Please raise more specific error types than plain `Exception`s. (See also L200)

lukpueh replied:
This one (and L200) are caught and re-raised with more information as specific `PacketParsingError`s in L254, adding additional information about the packet position.

aaaaalbert replied:
OK, got it. I wonder if it makes sense to raise the `PacketParsingError` on that level already, just so as to not catch other potential exceptions and report them in a confusing way.

(Then L254 would only `except PacketParsingError` and act upon it, no other unrelated exceptions.)

lukpueh replied:
Makes sense!

aaaaalbert commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+            PACKET_TYPE_USER_ATTR, PACKET_TYPE_SUB_KEY,
+            PACKET_TYPE_SIGNATURE]))
+
+    except Exception as e:
```
There's a bunch of `except Exception`s across this PR. I wonder why the blanket `except` is desired (instead of a more specific exception type), and why it makes sense in a few places to `continue` after logging the exception.

lukpueh replied:
Agreed! I do have this concern somewhere in my notes and I meant to add it to the PR description's "caveats/todos" section, as this issue applies to other places in the gpg module too, and I plan on addressing it in a separate PR. Thanks for pointing it out!

lukpueh replied:
Re `continue`: I considered that question too and decided that in some cases it is okay to continue the overall key export but ignore a self-certificate, if it
  - belongs to a malformed User ID packet (L307),
  - cannot be parsed  (L315),
  - is invalid (L337),

and, ignore a subkey, if it
  - cannot be parsed (L386, L389),
  - has a corresponding binding signature that cannot be parsed (L400),
  - has not exactly one binding signature (L420), or
  - the binding signature is invalid (L430).

But maybe it makes sense too do abort, since these errors are a sign that something is not fully supported or fishy.

lukpueh replied:
Will be addressed in https://github.com/in-toto/in-toto/issues/126#issuecomment-473337539.

aaaaalbert commented on in_toto/gpg/util.py:
```diff
@@ -100,34 +101,96 @@ def parse_packet_header(data, expected_type=None):
+
+  # If Bit 6 of 1st octet is set we parse a New Format Packet Length, and
+  # an Old Format Packet Lengths otherwise
+  if data[0] & 0x40: # pragma: no cover
```
Consider using binary representations for this and L132, `0b0100_0000` and `0b0011_1111` and such. (I think this is what they're here for :-)

lukpueh replied:
Great idea, will do.

lukpueh commented:
Thanks for your comments @aaaaalbert, they are very much appreciated! ❤️❤️❤️

Do you have any advice regarding the organization of  `get_pubkey_bundle `, `parse_pubkey_bundle`, `_assign_certified_key_info` and `_get_verified_subkeys`? I had this as one long spaghetti function initially, but that was too much to bear, so I split them into separate functions, but the boundaries feel a little arbitrary. Any thoughts?

aaaaalbert commented:
My pleasure! 😄 

------

Re organizing these functions, I think it's very sensible to have them separated even if they are called from a single place only. Each is moderately complicated, each has side- and sub-conditions in its checks, there must be tons of docs in each.... I'm totally fine with the way you did this!

The only thing that comes to mind is the choice of *where* exactly you collate the return values of the  private / sub-functions into what the caller uses. For example, [`parse_...` combines a few values](https://github.com/in-toto/in-toto/blob/d3b6c44fe76013568f3459f67ba2ed35db3935f0/in_toto/gpg/common.py#L262-L265); meanwhile [`get_...` could do the same right here](https://github.com/in-toto/in-toto/blob/d3b6c44fe76013568f3459f67ba2ed35db3935f0/in_toto/gpg/common.py#L493) with similar effort, assuming that it got `parse_...`'s `key_bundle` instead of the tuple it actually returns.

This isn't super-important, it's just to consider the time spent for someone who reads the code and has to trace function calls.

lukpueh commented:
@aaaaalbert, funny that you would mention this. Changing the code from the way you are proposing to what it is now was one of the last things I did on my WIP branch. 

The (probably only semi-compelling) reason why I'm calling the helper functions from within `parse_...` is that I didn't want to *expose* the raw key bundle (i.e. primary key, user id, user attribute, and sub key packets plus their respective signature packets) in order to not have it confused it with the key bundle (i.e. parsed primary key with certified additional info and parsed and bound subkeys) that's returned from `get_...` eventually.

But `parse_...` could of course be seen as just another helper function that returns a *private* data structure to be used by other helper functions.

aaaaalbert commented:
That's interesting! Any way is fine though, I suggest no change.

Commit:
Init primary key as empty dict to fix pylint error

Fix unsubscriptable-object error on Python2.7 that occurs due to a
dict operation on a variable that gets initialized as None first
and then must get re-initialized as dict inside a loop.

SantiagoTorres commented on in_toto/gpg/functions.py:
```diff
@@ -154,6 +157,11 @@ def gpg_verify_signature(signature_object, pubkey_info, content):
+            one of SHA1, SHA256, SHA512 (see in_toto.gpg.constants) used to
+            verify the signature. Default is SHA256.
+            NOTE: The hash algorithms specified in "pubkey_info"'s
+            "hashes" or "method" field are currently not considered.
```
I'm curious as to why we want to do this instead of updating our signature-type class?

lukpueh replied:
What do you mean by "signature-type class"? The information that's on the pubkey, i.e. [`hashes`](https://github.com/in-toto/in-toto/blob/v0.2.3/in_toto/gpg/formats.py#L33) and [`method`](https://github.com/in-toto/in-toto/blob/v0.2.3/in_toto/gpg/formats.py#L50)? If yes, than the reason is that that's the hashes/method we support for binary signatures, i.e. signatures over documents, which is the main purpose of the keys, and which still is only sha256.

I added sha1 and sha512 for self-signatures only, which,  IIRC, is something we agreed on in a different conversation. Does this make sense, or did I misunderstand your question?

SantiagoTorres commented on in_toto/gpg/util.py:
```diff
@@ -187,12 +250,12 @@ def parse_subpackets(subpacket_octets):
     ptr += 1
 
-    if length > 192 and length < 255 : # pragma: no cover
+    if length >= 192 and length < 255 : # pragma: no cover
```
Nice catch, thanks for correcting this.

aaaaalbert replied:
Minimal [style issue](https://github.com/secure-systems-lab/code-style-guidelines#pet-peeves): No leading space for the `:` here.

lukpueh replied:
@SantiagoTorres, it only got to my attention because I have a key in my system keyring, where that byte is exactly 192. :)
@aaaaalbert, thanks!

SantiagoTorres commented on in_toto/gpg/constants.py:
```diff
@@ -44,16 +44,18 @@
+PACKET_TYPE_SIGNATURE = 0x02
+PACKET_TYPE_PRIMARY_KEY = 0x06
+PACKET_TYPE_USER_ID = 0x0D
+PACKET_TYPE_USER_ATTR = 0x11
```
I'm assuming we need these two to concatenate them and verify the key-binding sig? If so, then ignore this. I just wanted to bring this up.

lukpueh replied:
You mean User ID and User Attribute? Yes, well no, not key-binding sigs but rather self-certs (0x10 - 0x13).  See [`_assign_certified_key_info()`](https://github.com/in-toto/in-toto/blob/5f18f390b0b706b0a6899f3ca6f050277570f355/in_toto/gpg/common.py#L268-L282) for more details.

lukpueh commented:
Thanks for your feedback, @SantiagoTorres and @aaaaalbert! I'll address your comments and add tests soon.

Commit:
Clean up gpg.util.get_hashing_class

Replace if cascade with dict/zip niceness. Kudos to @aaaaalbert
and @SantiagoTorres.

Commit:
Raise/catch specific error in parse_pubkey_bundle

Replace except generic Exception in parsing loop with actually
expected exceptions.

Commit:
Use binary notation for bitmasks in gpg.util

Use binary notation for bitmasks for better readability.

Commit:
Remove stray whitespace before colon

Commit:
Revert hash algo arg in gpg_verify_signature

Revert adding optional hash algorithm argument in
gpg.functions.gpg_verify_signature. This function is meant to
be used for binary type signatures, for which we only accept
sha265.

For other signature types (self sigs), where we accept other
hash algos too, we use a lower level gpg_verify_signature
function.

Commit:
Remove unused import in gpg.rsa

Commit:
Add gpg.util.get_hashing_class test

Commit:
Remove underscores in numeric binary literals

Only works for Python >=3.6
See https://www.python.org/dev/peps/pep-0515/#id25

Commit:
Change return values of gpg parse_packet_header

Change gpg.util.parse_packet_header to return the parsed
packet type and lengths instead of the body slice of the data
for easier testing.

Commit:
Mark unreachable code in parse_packet_header

Add comments and re-add "pragma: no cover's" to unreachable
code blocks in gpg.util.parse_packet.header

Commit:
Add gpg.util.parse_packet_header tests

Test packet header parsing with manually crafted packet data and
remove "pragma: no cover"s from newly tested code blocks.

Commit:
Reactor gpg.util.parse_subpackets

Factor out header parsing to new parse_subpacket_header function
for ease of testing and generally align implementation with the
corresponding pseudo-code in RFC4880.

Commit:
Add gpg.util.parse_subpacket_header tests

Commit:
Add gpg.common.parse_pubkey_payload tests

Test for raised exceptions with manually crafted pubkey data.

Note: Assert for exception message using context manager because
self.assertRaisesRegex is not available in Python2.

Commit:
Re-organize gpg pubkey parsing functions

Change parse_pubkey_bundle to return the parsed raw pubkey bundle
and let the caller, i.e. get_pubkey_bundle, call functions for
further processing, i.e. _assign_certified_key_info and
_get_verified_subkeys, to eventually return the processed
pubkey bundle, i.e. master_public_key (enriched with certified
key info and optional verified subkeys).

This follows suggestions made by @aaaaalbert in
https://github.com/in-toto/in-toto/pull/257#issuecomment-468303655
and also makes testing easier.

Commit:
Add new self-signature test key

Keyid is:
F557D0FF451DEF45372591429EA70BD13D883381

The test key has a 0x20 revocation signature on the primary key,
which, although we don't (yet) support key revocation signature handling,
can be used to test key bundle parsing.

The key also as an User Attribute packet with self-signature, which is
also parsed but not further handled.

Commit:
Add gpg.common._parse_pubkey_bundle tests

Add tests using manually crafted gpg data, partially based on
real gpg key data, to test error handling.

Commit:
Add gpg.common.get_pubkey_bundle test

Add test to trigger KeyNotFoundError

Commit:
Update gpg raw pubkey bundle data format

Also store the header length in raw gpg pubkey bundles for testing
purposes.

Commit:
Add gpg.common._parse_pubkey_bundle test

Add test that asserts the presence of packets expected returned
from `parse_pubkey_bundle` for specific test key). See
gpg --homedir tests/gpg_keyrings/rsa/ --export 9EA70BD13D883381 | \
   gpg --list-packets

Commit:
Update error handling in gpg pubkey parse helpers

Move signed content construction code out of try/except block.
If that codes fails something must have gone wrong earlier and
we should abort.

This commit also adds TODO notes to revise exception taxonomy.

Commit:
Add gpg.common._assign_certified_key_info tests

Use real gpg key data and modify it to trigger errors for testing.

Commit:
Add gpg.common._get_verified_subkeys tests

Use real gpg key data and modify it to trigger errors for testing.

Commit:
Add gpg.common.parse_signature_packet tests

Use manually crafted signature data to trigger errors for testing.

Commit:
Fix gpg.common tests for windows

Commit:
Exclude full gpg keyid related code from coverage

This differs across test environments.

lukpueh commented:
36085ad to 9a6141f are related to testing, including some minor modifications in the actual implementation  to make testing easier. Please review!

SantiagoTorres commented on in_toto/gpg/rsa.py:
```diff
@@ -26,7 +25,7 @@
 import in_toto.gpg.exceptions
 import in_toto.gpg.formats
-
+import in_toto.gpg.constants
```
I see that this removed one whitespace (we should have 2 here right?)

lukpueh replied:
Are you talking about blank lines after imports?

SantiagoTorres replied:
yes

lukpueh replied:
See bac9726

SantiagoTorres commented on in_toto/gpg/util.py:
```diff
@@ -26,6 +26,7 @@
 import in_toto.gpg.exceptions
 import in_toto.process
+import in_toto.gpg.constants
 
```
same whitespace issue.

lukpueh replied:
See bac9726

SantiagoTorres commented on in_toto/gpg/functions.py:
```diff
@@ -129,8 +131,8 @@ def gpg_sign_object(content, keyid=None, homedir=None):
-  # If we have a full keyid it is safe to remove the short keyid to save space
-  del signature["short_keyid"]
+  # It is okay now to remove the optional short keyid to save space
+  signature.pop("short_keyid", None)
```
Curious: what's the rationale for this change?

lukpueh replied:
`del signature["short_keyid"]` might raise a KeyError. Since `short_keyid` is optional, I'm using pop to be safe.

SantiagoTorres replied:
Ah right. Thanks for clarifying!

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+      # and it must be ordered because each signature packet belongs to the
+      # most recently parsed packet of a type.
+      elif packet_type in [PACKET_TYPE_USER_ID, PACKET_TYPE_USER_ATTR,
+          PACKET_TYPE_SUB_KEY]:
```
nit, can we use a set instead?

lukpueh replied:
sure will change

lukpueh replied:
Done a4a7e58

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+      # The remaining relevant packets are signatures, required to bind subkeys
+      # to the primary key, or to gather additional information about the
+      # primary key, e.g. expiration date.
+      # A signatures corresponds to the most recently parsed packet of a type,
```
s/signatures/signature/g

lukpueh replied:
Done 77d7fb1

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+
+    except Exception as e:
+      log.info(e)
+      continue
```
I wonder if we want to remove this `except` block and just let it blow up instead.

lukpueh replied:
I am doing this in a more recent commit.

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+
+      except Exception as e:
+        log.info(e)
+        continue
```
same here

lukpueh replied:
Here we shouldn't because `parse_signature_packet` might fail legitimately, because we want to continue if parse_signature_packet fails because we're giving it e.g. a revocation signature, which might happen. In a more recent commit I added a `TODO: Revise exception taxonomy`

SantiagoTorres commented on in_toto/gpg/dsa.py:
```diff
@@ -183,6 +183,8 @@ def gpg_verify_signature(signature_object, pubkey_info, content,
             one of SHA1, SHA256, SHA512 (see in_toto.gpg.constants) used to
             verify the signature
+            NOTE: Overrides any hash algorithm specification in "pubkey_info"'s
+            "hashes" or "method" fields.
```
I wonder if this note sounds more dramatic than it really is (as far as I understand `hash_algorithm_id` is populated with `pubkey_info["hashes"]` or `"method"` fields when not used directly?

lukpueh replied:
No. Currently `SHA256` is passed explicitly in the interface function `gpg.functions.gpg_verify_signature` to the corresponding dsa or rsa verification function. While it aligns with the possible values for `"hashes"` or `"method"`, assigned in `gpg.common.parse_pubkey_payload`, it is not populated from those fields.

But even if it were populated from those fields (IMO a good idea), we still want to be able to use a public key that we restrict to sha256 for binary signatures (i.e. `key_info["hashes"] = "pgp+SHA2"`), to verify self-signatures with other hash algos ... hence the support for an override and hence the comment. :)

We can of course re-think the fields in the public key dict and how we assign them. But I wouldn't do it as part of this PR.

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -329,14 +672,18 @@ def parse_signature_packet(data):
 
+    # info["subpackets"].append((
+    #   subpacket_type,
+    #   binascii.hexlify(subpacket_data).decode("ascii")))
```
This somehow sneaked through. Could you please remove it?

lukpueh replied:
I left it for michael, but I can remove it.

lukpueh replied:
I added a TODO comment 60555c7 and will remove or un-comment it with #245

SantiagoTorres reviewed (approved):
Whoa, this took me *hours* to review.

On the surface I could only find very minor stylistic stuff, nice job!

I'm approving this, and probably doing the style changes myself. I want to test backwards-compatibilty with old metadata, as I'm sure our integrators would appreciate.

Commit:
Fix blank lines at module head in gpg subpackage

Commit:
Use set instead of list in `in` condition

Addressing @SantiagoTorres' request:
https://github.com/in-toto/in-toto/pull/257#discussion_r267861031

Commit:
Fix typos in gpg subpackage comments

Commit:
Add TODO comment to commented out gpg block

SantiagoTorres commented:
LGTM.

Thanks @lukpueh !
